File: app/src/main/java/com/worldventures/dreamtrips/wallet/service/command/RecordListCommand.java
package com.worldventures.dreamtrips.wallet.service.command;

import com.innahema.collections.query.queriables.Queryable;
import com.worldventures.core.janet.dagger.InjectableAction;
import com.worldventures.dreamtrips.wallet.domain.entity.record.Record;
import com.worldventures.dreamtrips.wallet.service.SmartCardInteractor;
import com.worldventures.dreamtrips.wallet.service.command.wizard.DummyRecordCreator;
import com.worldventures.dreamtrips.wallet.util.WalletFeatureHelper;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import io.techery.janet.command.annotations.CommandAction;
import rx.functions.Func1;

@CommandAction
public class RecordListCommand extends CachedValueCommand<List<Record>> implements InjectableAction {

   @Inject WalletFeatureHelper featureHelper;
   @Inject SmartCardInteractor smartCardInteractor;

   public static RecordListCommand fetch() {
      return new RecordListCommand();
   }

   public static RecordListCommand remove(String recordId) {
      return new RecordListCommand(new RemoveOperationFunc(recordId));
   }

   public static RecordListCommand add(Record record) {
      return new RecordListCommand(new AddOperationFunc(record));
   }

   public static RecordListCommand addAll(List<Record> records) {
      return new RecordListCommand(new AddOperationFunc(records));
   }

   public static RecordListCommand replace(List<Record> records) {
      return new RecordListCommand(new ReplaceCardsOperationFunc(records));
   }

   public static RecordListCommand edit(Record record) {
      return new RecordListCommand(new EditOperationFunc(record));
   }

   public RecordListCommand() {
   }

   public RecordListCommand(Func1<List<Record>, List<Record>> operationFunc) {
      super(cards -> operationFunc.call(new ArrayList<>(cards)));
   }

   private static final class RemoveOperationFunc implements Func1<List<Record>, List<Record>> {

      private final String recordId;

      RemoveOperationFunc(String recordId) {
         this.recordId = recordId;
      }

      @Override
      public List<Record> call(List<Record> records) {
         records.remove(Queryable.from(records).first(element -> recordId.equals(element.id())));
         return records;
      }
   }

   private static final class AddOperationFunc implements Func1<List<Record>, List<Record>> {

      private final List<Record> recordsToAdd = new ArrayList<>();

      AddOperationFunc(Record record) {
         recordsToAdd.add(record);
      }

      AddOperationFunc(List<Record> records) {
         recordsToAdd.addAll(records);
      }

      @Override
      public List<Record> call(List<Record> records) {
         records.addAll(recordsToAdd);
         return records;
      }
   }

   private static final class ReplaceCardsOperationFunc implements Func1<List<Record>, List<Record>> {

      private final List<Record> newRecords = new ArrayList<>();

      ReplaceCardsOperationFunc(List<Record> records) {
         newRecords.addAll(records);
      }

      @Override
      public List<Record> call(List<Record> records) {
         records.clear();
         records.addAll(newRecords);
         return records;
      }
   }

   private static final class EditOperationFunc implements Func1<List<Record>, List<Record>> {

      private final Record editedRecord;

      EditOperationFunc(Record record) {
         editedRecord = record;
      }

      @Override
      public List<Record> call(List<Record> records) {
         return !records.isEmpty() ? Queryable.from(records).map(this::remapRecord).toList() : records;
      }

      private Record remapRecord(Record record) {
         return record.id().equals(editedRecord.id()) ? editedRecord : record;
      }
   }

   @Override
   protected void run(CommandCallback<List<Record>> callback) throws Throwable {
      if (featureHelper.isSampleCardMode()) {
         smartCardInteractor.smartCardUserPipe()
               .createObservableResult(SmartCardUserCommand.fetch())
               .map(command -> DummyRecordCreator.createRecords(command.getResult(), "0"))
               .subscribe(callback::onSuccess, callback::onFail);
      } else {
         super.run(callback);
      }
   }
}

File: wallet/src/main/java/com/worldventures/wallet/service/command/RecordListCommand.kt
package com.worldventures.wallet.service.command

import com.worldventures.janet.injection.InjectableAction
import com.worldventures.wallet.domain.entity.record.Record
import com.worldventures.wallet.service.SmartCardInteractor
import com.worldventures.wallet.service.command.wizard.DummyRecordCreator
import com.worldventures.wallet.util.WalletFeatureHelper
import io.techery.janet.command.annotations.CommandAction
import rx.functions.Func1
import java.util.ArrayList
import javax.inject.Inject

@CommandAction
class RecordListCommand : CachedValueCommand<List<Record>>, InjectableAction {

   @Inject lateinit var featureHelper: WalletFeatureHelper
   @Inject lateinit var smartCardInteractor: SmartCardInteractor

   private constructor()

   private constructor(func: Func1<MutableList<Record>, List<Record>>)
         : super(Func1<List<Record>, List<Record>> { func.call(it.toMutableList()) })

   private class RemoveOperationFunc internal constructor(private val recordId: String) : Func1<MutableList<Record>, List<Record>> {

      override fun call(records: MutableList<Record>): MutableList<Record> {
         records.remove(records.first { (id) -> recordId == id })
         return records
      }
   }

   private class AddOperationFunc : Func1<MutableList<Record>, List<Record>> {

      private val recordsToAdd = ArrayList<Record>()

      internal constructor(record: Record) {
         recordsToAdd.add(record)
      }

      internal constructor(records: MutableList<Record>) {
         recordsToAdd.addAll(records)
      }

      override fun call(records: MutableList<Record>): List<Record> {
         records.addAll(recordsToAdd)
         return records
      }
   }

   private class ReplaceCardsOperationFunc internal constructor(records: MutableList<Record>) : Func1<MutableList<Record>, List<Record>> {

      private val newRecords = ArrayList<Record>()

      init {
         newRecords.addAll(records)
      }

      override fun call(records: MutableList<Record>): List<Record> {
         records.clear()
         records.addAll(newRecords)
         return records
      }
   }

   private class EditOperationFunc internal constructor(private val editedRecord: Record) : Func1<MutableList<Record>, List<Record>> {

      override fun call(records: MutableList<Record>): List<Record> {

         return if (!records.isEmpty()) records.map { this.remapRecord(it) } else records
      }

      private fun remapRecord(record: Record): Record {
         return if (record.id.equals(editedRecord.id)) editedRecord else record
      }
   }

   @Throws(Throwable::class)
   override fun run(callback: CommandCallback<List<Record>>) {
      if (featureHelper.isSampleCardMode) {
         smartCardInteractor.smartCardUserPipe()
               .createObservableResult(SmartCardUserCommand.fetch())
               .map { command -> DummyRecordCreator.createRecords(command.result, "0") }
               .subscribe({ callback.onSuccess(it) }, { callback.onFail(it) })
      } else {
         super.run(callback)
      }
   }

   companion object {

      fun fetch(): RecordListCommand {
         return RecordListCommand()
      }

      fun remove(recordId: String): RecordListCommand {
         return RecordListCommand(RemoveOperationFunc(recordId))
      }

      fun add(record: Record): RecordListCommand {
         return RecordListCommand(AddOperationFunc(record))
      }

      fun addAll(records: MutableList<Record>): RecordListCommand {
         return RecordListCommand(AddOperationFunc(records))
      }

      fun replace(records: MutableList<Record>): RecordListCommand {
         return RecordListCommand(ReplaceCardsOperationFunc(records))
      }

      fun edit(record: Record): RecordListCommand {
         return RecordListCommand(EditOperationFunc(record))
      }
   }
}
